spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/DispatchingStreamListenerMessageHandler.java

```diff
@@ -0,0 +1,126 @@
+ * @author Marius Bogoevici
+ */
+final class DispatchingStreamListenerMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private final Collection<ConditionalStreamListenerHandler> handlerMethods;
+
+	private final EvaluationContext evaluationContext;
+
+	DispatchingStreamListenerMessageHandler(Collection<ConditionalStreamListenerHandler> handlerMethods,
+			EvaluationContext evaluationContext) {
+		Assert.notEmpty(handlerMethods, "'handlerMethods' cannot be empty");
+		Assert.notNull(evaluationContext, "'evaluationContext' cannot be empty");
+		this.handlerMethods = handlerMethods;
+		this.evaluationContext = evaluationContext;
+	}
+
+	@Override
+	protected boolean shouldCopyRequestHeaders() {
+		return false;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Collection<ConditionalStreamListenerHandler> matchingHandlers = findMatchingHandlers(requestMessage);
+		if (matchingHandlers.size() == 0) {
+			if (logger.isWarnEnabled()) {
+				logger.warn("Cannot find a @StreamListener matching " + requestMessage);
+			}
+			return null;
+		}
+		else if (matchingHandlers.size() > 1) {
+			for (ConditionalStreamListenerHandler matchingHandler : matchingHandlers) {
+				if (!matchingHandler.isVoid()) {
+					throw new MessagingException(
+							"Multiple matching methods cannot return values for " + requestMessage);
+				}
+			}
+			for (ConditionalStreamListenerHandler matchingMethod : matchingHandlers) {
+				matchingMethod.handleMessage(requestMessage);
+			}
+			return null;
+		}
+		else {
+			final ConditionalStreamListenerHandler singleMatchingHandler = matchingHandlers.iterator().next();
+			singleMatchingHandler.handleMessage(requestMessage);
+			return null;
+		}
+	}
+
+	private Collection<ConditionalStreamListenerHandler> findMatchingHandlers(Message<?> message) {
+		ArrayList<ConditionalStreamListenerHandler> matchingMethods = new ArrayList<>();
+		for (ConditionalStreamListenerHandler conditionalStreamListenerHandlerMethod : this.handlerMethods) {
+			if (conditionalStreamListenerHandlerMethod.getCondition() == null) {
+				matchingMethods.add(conditionalStreamListenerHandlerMethod);
+			}
+			else {
+				Object value = conditionalStreamListenerHandlerMethod.getCondition().getValue(this.evaluationContext, message);
+				if (Boolean.class.isAssignableFrom(value.getClass())) {
+					if ((Boolean) value) {
+						matchingMethods.add(conditionalStreamListenerHandlerMethod);
+					}
+				}
+				else if (Boolean.valueOf(String.valueOf(value))) {
+					matchingMethods.add(conditionalStreamListenerHandlerMethod);
+				}
+			}
+		}
+		return matchingMethods;
+	}
+
+	static class ConditionalStreamListenerHandler implements MessageHandler {
+
+		private Expression condition;
+
+		private StreamListenerMessageHandler streamListenerMessageHandler;
+
+		ConditionalStreamListenerHandler(Expression condition, StreamListenerMessageHandler streamListenerMessageHandler) {
```

Comment:
Would it be better to make `ConditionalStreamListenerHandler` with `non-null` condition value and have an intermediate base class to hold the `StreamListenerMessageHandler`. That way, instead of checking for `conditiion == null` for the non conditional stream listener handlers, we can have `type` check on `ConditionalStreamListenerHandler`.